FOCUS #366: SkuId, SkuPriceId, and Pricing-related columns review due to ChargeCategory and ChargeClass cleanup

ijurica opened the pull request:
ChargeCategory and ChargeClass cleanup impacted multiple columns. Based on current ChargeCategory references, the following columns must be revisited:

- SkuId
- SkuPriceId
- ContractedUnitPrice
- ListUnitPrice
- Pricingcategory
- PricingUnit
- PricingQuantity
- UsageQuantity
- UsageUnit
- ListCost
- ContractedCost
- EffectiveCost

ljadvey commented on supporting_content/columns/skupriceid.md:
```diff
@@ -25,15 +26,27 @@ Current values observed in billing data for various scenarios:
+
+| ChargeCategory | ChargeClass | perSku/bulk                       | SkuId            | SkuPriceId       |
+|----------------|-------------|-----------------------------------|------------------|------------------|
+| Usage          | Regular     | MUST be perSku and perSkuPrice    | MUST not be null | MUST not be null |
```
```suggestion
| Usage          | Original     | MUST be perSku and perSkuPrice    | MUST not be null | MUST not be null |
```

ljadvey replied:
Original change made on feedback from TF1 call

ljadvey commented on supporting_content/columns/skupriceid.md:
```diff
@@ -25,15 +26,27 @@ Current values observed in billing data for various scenarios:
+| Credit         | Correction  | MAY be bulk                       | MAY be null      | MAY be null      |
+| Adjustment     | Regular     | MAY be bulk                       | MAY be null      | MAY be null      |
+| Adjustment     | Correction  | MAY be bulk                       | MAY be null      | MAY be null      |
+| Tax            | Regular     | MUST be bulk                      | MUST be null     | MUST be null     |
```
```suggestion
| Tax            | Original     | MUST be bulk                      | MUST be null     | MUST be null     |
```

ljadvey commented on supporting_content/columns/skupriceid.md:
```diff
@@ -25,15 +26,27 @@ Current values observed in billing data for various scenarios:
+| Purchase       | Correction  | MAY be bulk                       | MAY be null      | MAY be null      |
+| Credit         | Regular     | MAY be bulk                       | MAY be null      | MAY be null      |
+| Credit         | Correction  | MAY be bulk                       | MAY be null      | MAY be null      |
+| Adjustment     | Regular     | MAY be bulk                       | MAY be null      | MAY be null      |
```
```suggestion
| Adjustment     | Original     | MAY be bulk                       | MAY be null      | MAY be null      |
```

ljadvey commented on supporting_content/columns/skupriceid.md:
```diff
@@ -25,15 +26,27 @@ Current values observed in billing data for various scenarios:
+| Usage          | Correction  | MAY be bulk                       | MAY be null      | MAY be null      |
+| Purchase       | Regular     | MUST be perSku and perSkuPrice    | MUST not be null | MUST not be null |
+| Purchase       | Correction  | MAY be bulk                       | MAY be null      | MAY be null      |
+| Credit         | Regular     | MAY be bulk                       | MAY be null      | MAY be null      |
```
```suggestion
| Credit         | Original     | MAY be bulk                       | MAY be null      | MAY be null      |
```

ljadvey commented on supporting_content/columns/skupriceid.md:
```diff
@@ -25,15 +26,27 @@ Current values observed in billing data for various scenarios:
+|----------------|-------------|-----------------------------------|------------------|------------------|
+| Usage          | Regular     | MUST be perSku and perSkuPrice    | MUST not be null | MUST not be null |
+| Usage          | Correction  | MAY be bulk                       | MAY be null      | MAY be null      |
+| Purchase       | Regular     | MUST be perSku and perSkuPrice    | MUST not be null | MUST not be null |
```
```suggestion
| Purchase       | Original     | MUST be perSku and perSkuPrice    | MUST not be null | MUST not be null |
```

ijurica commented:
### Changes overview

#### Adjusting normative paragraphs of multiple columns

The following table served as the basis for adjusting normative paragraphs of SkuPriceId spec, as well as price, cost, quantity metrics, etc., impacted by the ChargeCategory and ChargeClass columns cleanup

| ChargeCategory | ChargeClass | perSku/bulk                       | SkuId            | SkuPriceId       |
|----------------|-------------|-----------------------------------|------------------|------------------|
| Usage          | Regular     | MUST be perSku and perSkuPrice    | MUST not be null | MUST not be null |
| Usage          | Correction  | MAY be bulk                       | MAY be null      | MAY be null      |
| Purchase       | Regular     | MUST be perSku and perSkuPrice    | MUST not be null | MUST not be null |
| Purchase       | Correction  | MAY be bulk                       | MAY be null      | MAY be null      |
| Credit         | Regular     | MAY be bulk                       | MAY be null      | MAY be null      |
| Credit         | Correction  | MAY be bulk                       | MAY be null      | MAY be null      |
| Adjustment     | Regular     | MAY be bulk                       | MAY be null      | MAY be null      |
| Adjustment     | Correction  | MAY be bulk                       | MAY be null      | MAY be null      |
| Tax            | Regular     | MUST be bulk                      | MUST be null     | MUST be null     |
| Tax            | Correction  | MUST be bulk                      | MUST be null     | MUST be null     |

Same wording used in multiple columns:

- skuid.md
- skupriceid.md
- contractedunitprice.md
- listunitprice.md
- pricingcategory.md
- pricingquantity.md
- pricingunit.md
- usagequantity.md
- usageunit.md

**Example: PricingCategory**

> - PricingCategory MUST NOT be null when [ChargeClass](#chargeclass) is "Regular" and [ChargeCategory](#chargecategory) is "Usage" or "Purchase", MUST be null when ChargeCategory is "Tax", and MAY be null for all other combinations of ChargeClass and ChargeCategory.

#### Covering independent cost and/or pricing corrections

Note: Same wording used in multiple columns:

- contractedcost.md
- contractedunitprice.md
- listcost.md
- listunitprice.md
- pricingquantity.md

**Example: PricingQuantity**

> When unit prices are not null, multiplying PricingQuantity by a unit price MUST produce a result equal to the corresponding cost metric, **except in cases of ChargeClass "Correction", which MAY address PricingQuantity and/or any cost discrepancies independently**.

udam-f2 commented:
The group approved the change except for a minor update to lowercase the final 'MAY' -> may:
When unit prices are not null, multiplying PricingQuantity by a unit price MUST produce a result equal to the corresponding cost metric, except in cases of ChargeClass "Correction", which MAY address PricingQuantity and/or any cost discrepancies independently.

Also needs an update in changelog.

flanakin commented on specification/columns/contractedunitprice.md:
```diff
@@ -2,7 +2,7 @@
 The Contracted Unit Price represents the agreed-upon unit price for a single [Pricing Unit](#pricingunit) of the associated SKU, inclusive of negotiated discounts, if present, while excluding negotiated commitment-based discounts or any other discounts. This price is denominated in the [Billing Currency](#billingcurrency). The Contracted Unit Price is commonly used for calculating savings based o ...
 
-The ContractedUnitPrice column MUST be present in the billing data when the provider supports negotiated pricing concept. This column MUST be a Decimal within the range of non-negative decimal values, MUST conform to [Numeric Format](#numericformat) requirements, and be denominated in the BillingCurrency. When ContractedUnitPrice is present, it MUST NOT be null if [ChargeCategory](#chargecategory ...
+The ContractedUnitPrice column MUST be present in the billing data when the provider supports negotiated pricing concept. This column MUST be a Decimal within the range of non-negative decimal values, MUST conform to [Numeric Format](#numericformat) requirements, and be denominated in the BillingCurrency. It MUST NOT be null when [ChargeClass](#chargeclass) is "Regular" and [ChargeCategory](#char ...
```
```suggestion
The ContractedUnitPrice column MUST be present in the billing data when the provider supports negotiated pricing concept. This column MUST be a Decimal within the range of non-negative decimal values, MUST conform to [Numeric Format](#numericformat) requirements, and be denominated in the BillingCurrency. It MUST NOT be null when [ChargeClass](#chargeclass) is "Standard" and [ChargeCategory](#chargecategory) is "Usage" or "Purchase", MUST be null when ChargeCategory is "Tax", and MAY be null for all other combinations of ChargeClass and ChargeCategory. When ContractedUnitPrice is present and not null, multiplying ContractedUnitPrice by [PricingQuantity](#pricingquantity) MUST equal [ContractedCost](#contractedcost), except in cases of ChargeClass "Correction", which MAY address PricingQuantity and/or any cost discrepancies independently.
```

flanakin commented on specification/columns/listcost.md:
```diff
@@ -2,7 +2,7 @@
 List Cost represents the cost calculated by multiplying the [*list unit price*](#glossary:list-unit-price) and the corresponding [Pricing Quantity](#pricingquantity). List Cost is denominated in the [Billing Currency](#billingcurrency) and is commonly used for calculating savings based on various rate optimization activities, by comparing it with [Billed Cost](#billedcost) and [Effective Cost](#e ...
 
-The ListCost column MUST be present in the billing data and MUST NOT be null. This column MUST be of type Decimal, MUST conform to [Numeric Format](#numericformat) requirements, and be denominated in the BillingCurrency. When [ListUnitPrice](#listunitprice) is present and not null, multiplying the ListUnitPrice by PricingQuantity MUST produce the ListCost.
+The ListCost column MUST be present in the billing data and MUST NOT be null. This column MUST be of type Decimal, MUST conform to [Numeric Format](#numericformat) requirements, and be denominated in the BillingCurrency. When [ListUnitPrice](#listunitprice) is present and not null, multiplying the ListUnitPrice by PricingQuantity MUST produce the ListCost, except in cases of [ChargeClass](#charge ...
```
```suggestion
The ListCost column MUST be present in the billing data and MUST NOT be null. This column MUST be of type Decimal, MUST conform to [Numeric Format](#numericformat) requirements, and be denominated in the BillingCurrency. When [ListUnitPrice](#listunitprice) is present and not null, multiplying the ListUnitPrice by PricingQuantity MUST produce the ListCost, except in cases of [ChargeClass](#chargeclass) "Correction", which MAY address PricingQuantity or any cost discrepancies independently.
```

flanakin commented on specification/columns/listunitprice.md:
```diff
@@ -2,7 +2,7 @@
 The List Unit Price represents the suggested provider-published unit price for a single [Pricing Unit](#pricingunit) of the associated SKU, exclusive of any discounts. This price is denominated in the [Billing Currency](#billingcurrency). The List Unit Price is commonly used for calculating savings based on various rate optimization activities.
 
-The ListUnitPrice column MUST be present in the billing data when the provider publishes unit prices exclusive of discounts. This column MUST be a Decimal within the range of non-negative decimal values, MUST conform to [Numeric Format](#numericformat) requirements, and be denominated in the BillingCurrency. When ListUnitPrice is present, it MUST NOT be null if [ChargeCategory](#chargecategory) i ...
+The ListUnitPrice column MUST be present in the billing data when the provider publishes unit prices exclusive of discounts. This column MUST be a Decimal within the range of non-negative decimal values, MUST conform to [Numeric Format](#numericformat) requirements, and be denominated in the BillingCurrency. It MUST NOT be null when [ChargeClass](#chargeclass) is "Regular" and [ChargeCategory](#c ...
```
```suggestion
The ListUnitPrice column MUST be present in the billing data when the provider publishes unit prices exclusive of discounts. This column MUST be a Decimal within the range of non-negative decimal values, MUST conform to [Numeric Format](#numericformat) requirements, and be denominated in the BillingCurrency. It MUST NOT be null when [ChargeClass](#chargeclass) is "Standard" and [ChargeCategory](#chargecategory) is "Usage" or "Purchase", MUST be null when ChargeCategory is "Tax", and MAY be null for all other combinations of ChargeClass and ChargeCategory. When ListUnitPrice is present and is not null, multiplying ListUnitPrice by [PricingQuantity](#pricingquantity) MUST equal [ListCost](#listcost), except in cases of ChargeClass "Correction", which MAY address PricingQuantity and/or any cost discrepancies independently.
```

flanakin commented on specification/columns/pricingcategory.md:
```diff
@@ -5,7 +5,7 @@ Pricing Category describes the pricing model used for a charge at the time of us
 
 * PricingCategory MUST be present in the billing data when the provider supports more than one pricing category across all SKUs and MUST be of type String.
-* PricingCategory MUST NOT be null if [ChargeCategory](#chargecategory) is "Usage" or "Purchase" and MUST be null for other ChargeCategory values.
+* PricingCategory MUST NOT be null when [ChargeClass](#chargeclass) is "Regular" and [ChargeCategory](#chargecategory) is "Usage" or "Purchase", MUST be null when ChargeCategory is "Tax", and MAY be null for all other combinations of ChargeClass and ChargeCategory.
```
```suggestion
* PricingCategory MUST NOT be null when [ChargeClass](#chargeclass) is "Standard" and [ChargeCategory](#chargecategory) is "Usage" or "Purchase", MUST be null when ChargeCategory is "Tax", and MAY be null for all other combinations of ChargeClass and ChargeCategory.
```

jpradocueva commented:
Approved by the group on April 25, with changes. MAY to may and add changes to changelog
> PricingQuantity MUST produce the ListCost, except in cases of [ChargeClass](#chargeclass) "Correction", which `MAY` address PricingQuantity and/or any cost discrepancies independently.

flanakin commented on specification/columns/pricingquantity.md:
```diff
@@ -2,7 +2,7 @@
 The Pricing Quantity represents the volume of a given SKU associated with a [*resource*](#glossary:resource) or [*service*](#glossary:service) used or purchased, based on the [Pricing Unit](#pricingunit). Distinct from [Usage Quantity](#usagequantity) (complementary to [Usage Unit](#usageunit)), it focuses on pricing and cost, not *resource* and *service* consumption.
 
-The PricingQuantity column MUST be present in the billing data. This column MUST be of type Decimal and MUST conform to [Numeric Format](#numericformat) requirements. The value MAY be negative in cases where [ChargeClass](#chargeclass) is "Correction". This column MUST NOT be null if [ChargeCategory](#chargecategory) is "Usage" or "Purchase" and MAY be null for other ChargeCategory values. When u ...
+The PricingQuantity column MUST be present in the billing data. This column MUST be of type Decimal and MUST conform to [Numeric Format](#numericformat) requirements. The value MAY be negative in cases where [ChargeClass](#chargeclass) is "Correction". This column MUST NOT be null when [ChargeClass](#chargeclass) is "Regular" and [ChargeCategory](#chargecategory) is "Usage" or "Purchase", MUST be ...
```
```suggestion
The PricingQuantity column MUST be present in the billing data. This column MUST be of type Decimal and MUST conform to [Numeric Format](#numericformat) requirements. The value MAY be negative in cases where [ChargeClass](#chargeclass) is "Correction". This column MUST NOT be null when [ChargeClass](#chargeclass) is "Standard" and [ChargeCategory](#chargecategory) is "Usage" or "Purchase", MUST be null when ChargeCategory is "Tax", and MAY be null for all other combinations of ChargeClass and ChargeCategory. When unit prices are not null, multiplying PricingQuantity by a unit price MUST produce a result equal to the corresponding cost metric, except in cases of ChargeClass "Correction", which MAY address PricingQuantity or any cost discrepancies independently.
```

ijurica commented:
Suggested changes applied (changed 'MAY' to 'may'; replaced 'and/or' with 'or'; updated Changelog)
Ready for final review, approval, and hopefully merging.

jpradocueva commented:
@mike-finopsorg, @udam-f2 can one of you please approve this PR? It is ready to merge.
